File: selector/src/main/java/com/luck/picture/lib/config/SelectorProviders.java
package com.luck.picture.lib.config;

import java.util.LinkedList;

/**
 * @author：luck
 * @date：2023/3/31 4:15 下午
 * @describe：SelectorProviders
 */
public class SelectorProviders {

    private final LinkedList<SelectorConfig> selectionConfigsQueue = new LinkedList<>();

    public void addSelectorConfigQueue(SelectorConfig config) {
        selectionConfigsQueue.add(config);
    }

    public SelectorConfig getSelectorConfig() {
        return selectionConfigsQueue.size() > 0 ? selectionConfigsQueue.getLast() : new SelectorConfig();
    }

    public void destroy() {
        SelectorConfig selectorConfig = getSelectorConfig();
        if (selectorConfig != null) {
            selectorConfig.destroy();
            selectionConfigsQueue.remove(selectorConfig);
        }
    }

    public void reset() {
        for (int i = 0; i < selectionConfigsQueue.size(); i++) {
            SelectorConfig selectorConfig = selectionConfigsQueue.get(i);
            if (selectorConfig != null) {
                selectorConfig.destroy();
            }
        }
        selectionConfigsQueue.clear();
    }

    private static volatile SelectorProviders selectorProviders;

    public static SelectorProviders getInstance() {
        if (selectorProviders == null) {
            synchronized (SelectorProviders.class) {
                if (selectorProviders == null) {
                    selectorProviders = new SelectorProviders();
                }
            }
        }
        return selectorProviders;
    }
}

File: selector/src/main/java/com/luck/picture/lib/provider/SelectorProviders.kt
package com.luck.picture.lib.provider

import com.luck.picture.lib.config.SelectorConfig
import com.luck.picture.lib.utils.SelectorLogUtils
import java.util.*

/**
 * @author：luck
 * @date：2023/3/31 4:15 下午
 * @describe：SelectorProviders
 */
class SelectorProviders {
    private val configQueue = LinkedList<SelectorConfig>()

    fun addConfigQueue(config: SelectorConfig) {
        configQueue.add(config)
    }

    fun getConfig(): SelectorConfig {
        return if (configQueue.size > 0) configQueue.last else SelectorConfig()
    }

    fun destroy() {
        val config = getConfig()
        config.destroy()
        configQueue.remove(config)
        TempDataProvider.getInstance().reset()
        SelectorLogUtils.info("${System.currentTimeMillis()}:销毁")
    }

    fun reset() {
        for (i in configQueue.indices) {
            configQueue[i].destroy()
        }
        configQueue.clear()
    }

    companion object {
        fun getInstance() = InstanceHelper.instance
    }

    object InstanceHelper {
        val instance = SelectorProviders()
    }
}